File: cifras_agro/igac_gini/process/src/import_clean.jl
#
# Authors:     Sebas and Nicholas
# Maintainers: Sebas and Nicholas
# Year: 2024
# =========================================
# CEPAP/team_data/cifras_agro/igac_gini/input/GINI_MUNICIPAL_PERCENTILES.csv

using Pkg, DataFrames, CSV, Random, Dates, StringEncodings, Transliterate, TidierFiles

Random.seed!(19481210)

println("load functions--", now())
replace_commas(x) = [ismissing(item) ? missing : replace(item, "," => ".") for item in x]
coerce_deci(x) = [ismissing(item) ? missing : parse(Float64, item) for item in x]

println("load data--", now())
input_1 = "C:/Users/sebas/OneDrive/Documents/CEPAP/team_data/cifras_agro/igac_gini/process/input/GINI_MUNICIPAL_PERCENTILES.csv"
gini_igac = DataFrame(CSV.File(open(input_1, enc"ISO-8859-1")))

println("clean column names--", now())
rename!(gini_igac, Symbol.(replace.(lowercase.(names(gini_igac)), r"\s+|\.+|__+" => "_")))
rename!(gini_igac, replace.(names(gini_igac), "__" => "_"))

println("count missing values--", now())
n_missng_gini = count(ismissing, gini_igac.gini)
n_missng_area_has = count(ismissing, gini_igac.area_ha)

println("clean values--", now())
string_cols = names(gini_igac, String)
transform!(gini_igac, string_cols .=> ByRow(transliterate) .=> string_cols)
cols_numrc = ["area_ha", "gini"]
transform!(gini_igac, cols_numrc .=> (col -> coerce_deci(replace_commas(col))) .=> cols_numrc)
@assert count(ismissing, gini_igac.gini) == n_missng_gini && count(ismissing, gini_igac.area_ha) == n_missng_area_has

println("filter--", now())
gini_igac_perc = gini_igac[.!ismissing.(gini_igac.percentiles), :]
@assert all(combine(groupby(gini_igac_perc, :cod_muni), :cod_muni => length => :n_recs).n_recs .== 300)
gini_igac = gini_igac[ismissing.(gini_igac.percentiles), :]
@assert all(combine(groupby(gini_igac, :cod_muni), :cod_muni => length => :n_recs).n_recs .== 3)

println("select cols--", now())
n_cols_gini_igac_perc = ncol(gini_igac_perc)
n_col_igac = ncol(gini_igac)
vars_selct = setdiff(names(gini_igac), ["no_propietarios", "no_predios", "area_ha",
"area_m2", "disparidad_superior"])
select!(gini_igac_perc, vars_selct)
@assert ncol(gini_igac_perc) == n_cols_gini_igac_perc - 5
vars_selct = setdiff(names(gini_igac), ["percentiles"])
select!(gini_igac, vars_selct)
@assert ncol(gini_igac) == n_col_igac - 1

println("export--", now())
output_1 = "C:/Users/sebas/OneDrive/Documents/CEPAP/team_data/cifras_agro/igac_gini/process/output/gini_igac.parquet"
output_2 = "C:/Users/sebas/OneDrive/Documents/CEPAP/team_data/cifras_agro/igac_gini/process/output/gini_igac_perc.parquet"
write_parquet(gini_igac, output_1)
write_parquet(gini_igac_perc, output_2)
gini_igac = nothing
gini_igac_perc = nothing

println("done import_clean.jl--", now())

File: cifras_agro/igac_gini/process/src/import_clean.R
#
# Authors:     Sebas and Nicholas
# Maintainers: Sebas and Nicholas
# Year: 2024
# =========================================
# CEPAP/team_data/cifras_agro/igac_gini/input/GINI_MUNICIPAL_PERCENTILES.csv

pacman::p_load(argparse, here, dplyr, arrow, assertr, logger, readr, janitor,
               stringi, stringr)

set.seed(19481210)

if(dir.exists("/Users/sebas/OneDrive")) {
  
  input_1 <- "/Users/sebas/OneDrive/Documents/CEPAP/team_data/cifras_agro/igac_gini/process/input/GINI_MUNICIPAL_PERCENTILES.csv"
  
  output_1 <- "/Users/sebas/OneDrive/Documents/CEPAP/team_data/cifras_agro/igac_gini/process/output/gini_igac.parquet"
  
  output_2 <- "/Users/sebas/OneDrive/Documents/CEPAP/team_data/cifras_agro/igac_gini/process/output/gini_igac_perc.parquet"
} else {
  input_1 <- "/Users/ruta_nico"
}

log_info("load data")
gini_igac <- read_delim(input_1, delim = ";", locale = locale(encoding = "ISO-8859-1"))

log_info("clean")
gini_igac <- gini_igac %>% 
  clean_names()

n_na_gini <- sum(is.na(gini_igac$gini))

gini_igac <- gini_igac %>% 
  mutate_if(is.character, ~stri_trans_general(., "latin-ascii")) %>% 
  mutate(gini = str_replace_all(gini, ",", ".")) %>% 
  mutate(gini = as.numeric(gini)) %>% 
  verify(sum(is.na(gini)) == n_na_gini)

log_info("filter")
gini_igac_perc <- gini_igac %>% 
  filter(!(is.na(percentiles)))

gini_igac_perc %>% 
  group_by(cod_muni) %>% 
  summarise(n_recs = n()) %>% 
  ungroup() %>% 
  verify(all(n_recs == 300))

gini_igac <- gini_igac %>% 
  filter(is.na(percentiles))

gini_igac %>% 
  group_by(cod_muni) %>% 
  summarise(n_recs = n()) %>% 
  ungroup() %>% 
  verify(all(n_recs == 3))

log_info("select cols")
n_cols_gini_igac_per <- ncol(gini_igac_perc)

n_col_igac <- ncol(gini_igac)

gini_igac_perc <- gini_igac_perc %>% 
  select(-no_propietarios, -no_predios, -area_ha, -area_m2, -disparidad_superior) %>% 
  distinct() %>% 
  verify(ncol(.) == n_cols_gini_igac_per - 5)
  
gini_igac <- gini_igac %>% 
  select(-percentiles) %>% 
  distinct() %>% 
  verify(ncol(.) == n_col_igac - 1)

log_info("export")
write_parquet(gini_igac, output_1)

write_parquet(gini_igac_perc, output_2)

rm(gini_igac, gini_igac_perc)
gc()

log_info("done clean.R")